### ISSUE TITLE
make it easier to lint the code

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#2818]
When working with Python at work (my projects primary language) it's mandatory that the code be lint free; i.e., pylint produces no diagnostics. This has been extremely useful both because pylint sometimes catches real bugs and because I know if I lint my local repo and it isn't lint free I've introduced lint. For those reasons the fish code should also be lint clean. The first step in doing so is to make it easy for people who aren't using something like the OS X Xcode IDE to lint the code. To achieve that I propose adding `lint` and `lint-all` *make* targets. The next step, which we'll tackle in a separate issue, is to deal with all the existing lint diagnostics.

I started work today on a script that can be invoked from the *Makefile* that will run the code through cppcheck and oclint. The `make lint` command will lint just those files being changed that are not yet committed. If there are none then it will lint the files in the most recently committed change. The `make lint-all` command will, as the name implies, lint all the source in the *src* directory.

**Comment by krader1961** [comment]
I should also mention that per an earlier conversation with @ridiculousfish I'm planning on supporting cppcheck and oclint. I don't think we want to support more than two lint tools and would prefer only one; at least as authoritative arbiters of lint. We can always add support for other tools that are not authoritative if people find them personally useful.

**Comment by floam** [comment]
It seems a little strange Xcode has all these FANTASTIC built-intools, but actually unlike the Makefile has no tests defined. It'd be nice if the `make test` tests could be snuck in as some kind of post-build process, and perhaps a smattering of XCTess could do something unique? 

I use OCLint and have it integrated into Xcode. I chose to use [xctool](https://github.com/facebook/xctool) to make it happen and it's actually quite nice, I'd never hearad of it before. Among other stuff one can have xcode outputting "Test Anything Protocol", which could actually rope linting, unit testing, static and runtime analysis together with a common interface on under the  Make, travis, and xcode roofs.

**Comment by krader1961** [comment]
Thanks, @floam, for the feedback. Adding support for platform specific tools like xcodebuild and xctool is definitely a good idea but is probably a stretch goal that should be deferred until we've got basic CLI support that works on both OS X and LInux. Once that is done people like yourself can augment the implementation to make linting easier for humans to deal with and more comprehensive on a specific platform.

My primary goal with this issue is to implement a simple way for anyone creating a pull-request, whether they're using OS X or Linux, to know if they've introduced any obvious bogosities via a `make lint` invocation.

**Comment by zanchey** [comment]
It would be good if that can be added to the Travis builds, as well - because at least then if you don't have the tools installed then a pull request can be tested.

I had vague plans of running `make iwyu` against the tree in Travis but haven't gotten around to it yet.

**Comment by krader1961** [comment]
> It would be good if that can be added to the Travis builds...

I agree. That is the logical next step but can only be done once we agree being "lint clean" is desirable and the code is in fact lint clean. That is true for my primary project at $dayjob and I think it is a useful goal for this project.

**Comment by floam** [comment]
I think in order to be capable of agreeing it would help to have an idea of how much work is necessary on the tree now, what kinds of changes that may be expected in the future, and how valuable the feedback provided by these tools is. Do you have a set of proposed settings for these tools and the build system as to what would define Lint Clean™, so people could try that out, or some examples?

**Comment by krader1961** [comment]
> Do you have a set of proposed settings for these tools...

Not yet. I'm invoking both tools with their generic settings at this point. I have a local branch where I'm running `make lint-all`. It identifies so many problems in the case of oclint that it refuses to report more errors. The very first problem it reports appears to be a real bug. Cppcheck is reporting a much smaller number of problems but all of which appear to be legitimate areas of concern.

Lint tools are usually conservative and when they aren't have options for suppressing warnings that are questionable (which my $dayjob project takes advantage of).

**Comment by krader1961** [comment]
Grumble. I'm having a really hard time getting ccpcheck and oclint to work on OS X. I'm getting lots of nonsensical errors about #include's not being found without adding hardcoded **-I** arguments to their invocations. So earlier today I decided to switch to working on Linux. Doing so is producing more sensible behavior (i.e., I don't need to include options that are not relevant to the project).

The reason for mentioning this is that I still feel quite strongly that being able to lint changes via something like `make lint` should be easy to do. However, making that work in practice is proving more difficult than I expected.

P.S., The code has so many lint errors, at least a few of which seem to point to actual bugs, that it, is going to take a lot of effort to make the project lint clean.

**Comment by floam** [comment]
```cppcheck --enable=all --inconclusive --std=posix fish-shell/src -v``` works fine for me on OS X.

**Comment by floam** [comment]
if you have pygments:
```
$ cppcheck --enable=all --inconclusive --std=posix -v --xml --xml-version=2 fish-shell/src/ 2> cppcheck.xml
$ cppcheck-htmlreport --title fish-HEAD --file=./cppcheck.xml - --report-dir=.
Parsing xml report.
Creating . directory
Processing errors
  fish-shell/src/autoload.cpp
  fish-shell/src/builtin.cpp
  fish-shell/src/builtin_commandline.cpp
  fish-shell/src/builtin_complete.cpp
  fish-shell/src/builtin_jobs.cpp
  fish-shell/src/builtin_printf.cpp
  fish-shell/src/builtin_set.cpp
  fish-shell/src/builtin_set_color.cpp
  fish-shell/src/builtin_string.cpp
  fish-shell/src/builtin_test.cpp
  fish-shell/src/builtin_ulimit.cpp
  fish-shell/src/color.cpp
  fish-shell/src/common.cpp
  fish-shell/src/common.h
  fish-shell/src/complete.cpp
  fish-shell/src/env.cpp
  fish-shell/src/env.h
  fish-shell/src/env_universal_common.cpp
  fish-shell/src/event.cpp
  fish-shell/src/exec.cpp
  fish-shell/src/expand.cpp
  fish-shell/src/fallback.cpp
  fish-shell/src/fish.cpp
  fish-shell/src/fish_indent.cpp
  fish-shell/src/fish_tests.cpp
  fish-shell/src/fish_version.cpp
  fish-shell/src/function.cpp
  fish-shell/src/highlight.cpp
  fish-shell/src/history.cpp
  fish-shell/src/input.cpp
  fish-shell/src/input_common.cpp
  fish-shell/src/intern.cpp
  fish-shell/src/io.cpp
  fish-shell/src/iothread.cpp
  fish-shell/src/key_reader.cpp
  fish-shell/src/kill.cpp
  fish-shell/src/lru.h
  fish-shell/src/output.cpp
  fish-shell/src/pager.cpp
  fish-shell/src/parse_execution.cpp
  fish-shell/src/parse_productions.cpp
  fish-shell/src/parse_tree.cpp
  fish-shell/src/parse_tree.h
  fish-shell/src/parse_util.cpp
  fish-shell/src/parser.cpp
  fish-shell/src/parser_keywords.cpp
  fish-shell/src/path.cpp
  fish-shell/src/postfork.cpp
  fish-shell/src/proc.cpp
  fish-shell/src/reader.cpp
  fish-shell/src/sanity.cpp
  fish-shell/src/screen.cpp
  fish-shell/src/screen.h
  fish-shell/src/signal.cpp
  fish-shell/src/tokenizer.cpp
  fish-shell/src/utf8.cpp
  fish-shell/src/util.cpp
  fish-shell/src/wcstringutil.cpp
  fish-shell/src/wgetopt.cpp
  fish-shell/src/wildcard.cpp
  fish-shell/src/wutil.cpp
Creating index.html
Creating style.css file
Creating stats.html (statistics)


[Open './index.html' to see the results.](http://aaron.gy/cppcheck.html)
```
